File: apps/mobile/lib/features/auth/data/auth_adapter.dart
import 'package:psiops_contracts/api.dart';

/// Porta de acesso à autenticação da psicóloga.
///
/// Os tipos [AuthResponse] e [TokenPair] vêm de `packages/contracts/gen/dart`
/// (codegen do `openapi.yaml` — fonte única de contratos, ADR 0008); este
/// adapter nunca redefine DTOs de API localmente.
///
/// Duas implementações:
/// - `InMemoryAuthAdapter`: mock em memória, padrão em dev/test.
/// - `HttpAuthAdapter`: client HTTP real tipado pelos mesmos modelos,
///   implementado e compilável, mas não exercitado contra a API real nesta
///   tarefa (integração real é PSI-045).
///
/// A escolha entre as duas acontece em um único ponto de composição:
/// `app/app.dart` (`_authAdapterFor`).
abstract interface class AuthAdapter {
  /// Autentica com e-mail e senha (`POST /auth/login`).
  ///
  /// Lança [InvalidCredentialsException] em credenciais inválidas.
  Future<AuthResponse> login(String email, String password);

  /// Registra uma nova conta e já inicia a sessão (`POST /auth/register`).
  ///
  /// Lança [EmailAlreadyRegisteredException] se o e-mail já estiver em uso.
  Future<AuthResponse> register(String name, String email, String password);

  /// Troca um refresh token válido por um novo par de tokens
  /// (`POST /auth/refresh`). O refresh token apresentado é invalidado
  /// (rotação de uso único), conforme a spec.
  ///
  /// Lança [RefreshFailedException] em falha definitiva (refresh token
  /// inválido, expirado ou já utilizado) — quem chama deve encerrar a sessão.
  Future<TokenPair> refresh(String refreshToken);
}

/// Erro genérico de autenticação (rede, servidor, resposta inesperada).
///
/// [message] é um texto pt-BR adequado para exibição direta na UI.
class AuthAdapterException implements Exception {
  const AuthAdapterException(this.message);

  final String message;

  @override
  String toString() => 'AuthAdapterException: $message';
}

/// Credenciais inválidas em `/auth/login` (HTTP 401).
class InvalidCredentialsException extends AuthAdapterException {
  const InvalidCredentialsException([
    super.message = 'E-mail ou senha incorretos.',
  ]);
}

/// E-mail já cadastrado em `/auth/register` (HTTP 409).
class EmailAlreadyRegisteredException extends AuthAdapterException {
  const EmailAlreadyRegisteredException([
    super.message = 'Este e-mail já está cadastrado.',
  ]);
}

/// Falha definitiva ao renovar o par de tokens em `/auth/refresh` (HTTP 401,
/// ou ausência de refresh token) — a sessão deve ser encerrada e a usuária
/// redirecionada ao login com esta mensagem.
class RefreshFailedException extends AuthAdapterException {
  const RefreshFailedException([
    super.message = 'Sua sessão expirou. Faça login novamente.',
  ]);
}

File: apps/clinic/src/adapters/auth/AuthAdapter.ts
import type {
  AuthResponse,
  LoginRequest,
  RefreshTokenRequest,
  RegisterRequest,
  SessionResponse,
  TokenPair,
} from "@psiops/contracts";

/**
 * Interface de acesso à autenticação (ADR 0006 — frontends desacoplados por
 * adapters). Tipada exclusivamente pelos tipos gerados de `@psiops/contracts`
 * (`gen/ts`), espelhando 1:1 os quatro endpoints de `/auth/*`; nenhum DTO é
 * redefinido aqui.
 *
 * Implementações:
 * - `MockAuthAdapter` — estado em memória, determinístico, padrão em
 *   desenvolvimento e testes (usuário semente, simulação de expiração).
 * - `HttpAuthAdapter` — tipada contra os contratos, aponta para a API Spring
 *   Boot, mas sem chamadas reais habilitadas nesta tarefa (integração
 *   completa fica para a PSI-044).
 *
 * O ponto de composição único (seleção mock/http por variável de ambiente)
 * fica em `./index.ts`.
 */
export interface AuthAdapter {
  /** `POST /auth/login` — credenciais inválidas rejeita com `AuthError` (status 401). */
  login(payload: LoginRequest): Promise<AuthResponse>;

  /** `POST /auth/register` — e-mail já cadastrado rejeita com `AuthError` (status 409). */
  register(payload: RegisterRequest): Promise<AuthResponse>;

  /**
   * `POST /auth/refresh` — troca um refresh token válido por um novo par
   * (rotação de uso único). Refresh token inválido/expirado/já usado rejeita
   * com `AuthError` (status 401).
   */
  refresh(payload: RefreshTokenRequest): Promise<TokenPair>;

  /**
   * `GET /auth/session` — retorna a sessão corrente para o access token
   * apresentado. Token ausente/expirado/revogado rejeita com `AuthError`
   * (status 401). É a operação usada para exercitar o refresh transparente
   * (ver `SessionManager.withAuth`).
   */
  getSession(accessToken: string): Promise<SessionResponse>;
}
